services/user/Workshop/ui/src/components/metadata-form.tsx

```diff
@@ -0,0 +1,271 @@
+});
+
+export type Schema = z.infer<typeof formSchema>;
+
+interface Props {
+  existingValues?: Schema;
+  onSubmit: (data: Schema) => Promise<Schema>;
+}
+
+const blankDefaultValues = formSchema.parse({
+  name: "",
+  shortDescription: "",
+  longDescription: "",
+  icon: "",
+  iconMimeType: "",
+  tosSubpage: "",
+  privacyPolicySubpage: "",
+  appHomepageSubpage: "",
+  redirectUris: [],
+  tags: [],
+});
+
+export const MetaDataForm = ({ existingValues, onSubmit }: Props) => {
+  const form = useForm({
+    resolver: zodResolver(formSchema),
+    defaultValues: existingValues || blankDefaultValues,
+  });
+
+  const [iconPreview, setIconPreview] = useState<string | null>(null);
+
+  const submit = async (data: Schema) => {
+    try {
+      form.clearErrors();
+      toast("Saving...");
+      await onSubmit({
+        ...data,
+        tosSubpage: data.tosSubpage || "/",
+        appHomepageSubpage: data.appHomepageSubpage || "/",
+        privacyPolicySubpage: data.privacyPolicySubpage || "/",
+      });
+      form.reset(data);
+      toast("Success", { description: `Saved changes.` });
+    } catch (error) {
+      if (error instanceof ZodError) {
+        error.issues.forEach((field) => {
+          form.setError(field.path[0] as keyof Schema, {
+            message: field.message,
+          });
+        });
+        toast("Invalid submssion", {
+          description: "One or more fields are invalid.",
+        });
+      } else if (error instanceof Error) {
+        form.setError("root", {
+          message:
+            error instanceof Error
+              ? error.message
+              : "Unrecognised error, see log.",
+        });
+        toast("Request failed", { description: error.message });
+      }
+    }
+  };
+
+  const isIcon =
+    iconPreview ||
+    (existingValues && existingValues.icon && existingValues.iconMimeType);
+  const currentSrc = `data:${existingValues?.iconMimeType};base64,${existingValues?.icon}`;
+
+  const handleIconChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
+    const file = e.target.files?.[0];
+    if (file) {
+      const mimeType = file.type;
+      const base64Icon = await fileToBase64(file);
+
+      const iconSrc = `data:${mimeType};base64,${base64Icon}`;
+      form.setValue("icon", base64Icon);
+      form.setValue("iconMimeType", mimeType);
+      setIconPreview(iconSrc);
+
+      const reader = new FileReader();
+      reader.readAsDataURL(file);
+      reader.onloadend = () => {
+        setIconPreview(reader.result as string);
+      };
+    }
+  };
+
+  return (
+    <Form {...form}>
+      <form
+        onSubmit={form.handleSubmit(submit)}
+        className="flex flex-col gap-2"
+      >
+        <FormField
+          control={form.control}
+          name="name"
+          render={({ field }) => (
+            <FormItem>
+              <FormLabel>App Name</FormLabel>
+              <FormControl>
+                <Input placeholder="MonsterApp" {...field} />
```

Comment:
I'd recommend a different default example here instead of a game. Maybe something related to voting or trading. Or a prediction market of some kind. 🤔